### PR TITLE
ATS-702 Add AIOControllerTest for Imagemagick transforms

### PR DESCRIPTION
**Pull request by PunchedChimera**
Using the class inheritance already established, the tests executed by the ImageMagickControllerTest class are now run against the new AIOController.

alfresco-transform-imagemagick-boot now produces a test-jar that is utilized by alfresco-transform-core-aio-boot in order to achieve this.

Fixes have been applied, following failing builds.

### CODE REVIEW

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerImageMagickTest.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            .param("sourceMimetype", sourceMimetype);
+
+        return builder;
+    }
+
+    @Test
+    @Override
+    public void noTargetFileTest()
+    {
+        // Ignore the test in super class as the AIO transforms we not be selected .
```
Comment (montgolfiere) [inline alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerImageMagickTest.java:110]: Trivial: grammar -> "we not be selected ." -> "was not selected." ? Similar below ...

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerHttpRequestTest.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/*
+ * #%L
+ * Alfresco Transform Core
+ * %%
+ * Copyright (C) 2005 - 2019 Alfresco Software Limited
```
Comment (montgolfiere) [inline alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerHttpRequestTest.java:5]: Trivial: 2020 ?

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/main/java/org/alfresco/transformer/AIOController.java [hunk 3]
```diff
@@ -175,16 +177,8 @@ protected void executeTransformCommand(File sourceFile, File targetFile)
         // TODO - remove this logginng
```
Comment (montgolfiere) [inline alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/main/java/org/alfresco/transformer/AIOController.java:177]: Resolve this TODO ?

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/resources/engine_config_complete.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
+{
```
Comment (montgolfiere) [inline alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/resources/engine_config_complete.json:1]: Is this a duplicate test file or new ? Similar for all other test resource files below ... ?

If duplicate then maybe raise tech debt to rationalise in future ?

File: alfresco-transform-tika/alfresco-transform-tika/pom.xml [hunk 5]
```diff
@@ -49,6 +49,10 @@
                     <groupId>org.quartz-scheduler</groupId>
                     <artifactId>quartz</artifactId>
                 </exclusion>
+                <exclusion>
```
Comment (montgolfiere) [inline alfresco-transform-tika/alfresco-transform-tika/pom.xml:52]: Should this be a separate PR - does not seem related to ImageMagick tests. See ATS-692 ?

**Comment by montgolfiere** [comment]
Note: as agreed PR # 200 (branch ATS-675_aio_transformer) has now been merged to master.

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/main/java/org/alfresco/transformer/AIOController.java [hunk 6]
```diff
@@ -118,9 +118,9 @@ protected void executeTransformCommand(File sourceFile, File targetFile)
     @PostMapping(value = "/transform", consumes = MULTIPART_FORM_DATA_VALUE)
     public ResponseEntity<Resource> transform(HttpServletRequest request,
         @RequestParam("file") MultipartFile sourceMultipartFile,
+        @RequestParam(TARGET_EXTENSION) String targetExtension,
```
Comment (PunchedChimera) [inline alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/main/java/org/alfresco/transformer/AIOController.java:121]: Existing tests expect this parameter to be checked first.

File: alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/src/test/java/org/alfresco/transformer/ImageMagickControllerTest.java [hunk 7]
```diff
@@ -26,77 +26,23 @@
...
-    @Mock
-    private RuntimeExec mockTransformCommand;
-
-    @Mock
-    private RuntimeExec mockCheckCommand;
-
-    private ImageMagickCommandExecutor commandExecutor = new ImageMagickCommandExecutor();
 
     @SpyBean
     private ImageMagickController controller;
```
Comment (PunchedChimera) [inline alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/src/test/java/org/alfresco/transformer/ImageMagickControllerTest.java:48]: Separating the ImageMagick tests were needed as this inherited variable could not be instantiated by AIO test classes.

File: alfresco-transformer-base/src/test/java/org/alfresco/transformer/AbstractTransformerControllerTest.java [hunk 8]
```diff
@@ -163,14 +169,35 @@ void generateTargetFileFromResourceFile(String actualTargetExtension, File testF
...
+                {
+                    testFile = new File(URLDecoder.decode(folder.getRoot().getPath()+ File.separator + testFilename, "UTF-8"));
+                }
+                else
+                {
+                    throw e;
+                }
+            }
+        }
+
```
Comment (PunchedChimera) [inline alfresco-transformer-base/src/test/java/org/alfresco/transformer/AbstractTransformerControllerTest.java:199]: Allows files to be inherited test files to be read from the jars, so that duplicated resources are not required.

**Comment by montgolfiere** [comment]
@PunchedChimera If you update branch from master then should fix/remove red SAST job.

File: alfresco-transformer-base/src/test/java/org/alfresco/transformer/AbstractTransformerControllerTest.java (hunk 8 above)
Comment (montgolfiere) [reply alfresco-transformer-base/src/test/java/org/alfresco/transformer/AbstractTransformerControllerTest.java:199]: Maybe add this reason as a comment to the code above ?

File: alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/src/test/java/org/alfresco/transformer/ImageMagickControllerTest.java (hunk 7 above)
Comment (montgolfiere) [reply alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/src/test/java/org/alfresco/transformer/ImageMagickControllerTest.java:48]: Maybe add this reason as a comment to code ?

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/main/java/org/alfresco/transformer/AIOController.java (hunk 6 above)
Comment (PunchedChimera) [reply alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/main/java/org/alfresco/transformer/AIOController.java:121]: It's checked as a negative test. This is the minimum parameter required across all T-engines and the Super test class AbstractTransformerControllerTest only checks for an error on the TARGET_EXTENSION parameter.  If any other parameters are run before (and not included) it does not present the expected errror.

File: alfresco-transform-core-aio/alfresco-transform-core-aio/src/test/java/org/alfresco/transformer/transformers/AllTransformRegistryTest.java [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+        Set<String> actualOptionNames = aioTransformer.getTransformConfig().getTransformOptions().keySet();
+
+        // check all options are there
+        for (String optionName : expectedTransformOptionNames)
+        {
+            assertTrue("Expected transform option missing.",  actualOptionNames.contains(optionName));
+        }
+    }
+
+    /// Test copied from Misc aioTransformer - html
```
[the remark this answers is not in the document]
Comment (PunchedChimera) [reply alfresco-transform-core-aio/alfresco-transform-core-aio/src/test/java/org/alfresco/transformer/transformers/AllTransformRegistryTest.java:130]: Actually a duplicated file (now removed) of AIOTransformRegistryTest from master, which include JIRA references

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerImageMagickTest.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+    }
+
+    @Before
+    public void before() throws IOException, Exception
+    {
+        adapter = new ImageMagickAdapter();
+        ReflectionTestUtils.setField(commandExecutor, "transformCommand", mockTransformCommand);
+        ReflectionTestUtils.setField(commandExecutor, "checkCommand", mockCheckCommand);
+        ReflectionTestUtils.setField(adapter, "commandExecutor", commandExecutor);
+        //Need to wire in the mocked adpater into the controller...
```
[the remark this answers is not in the document]
Comment (PunchedChimera) [reply alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerImageMagickTest.java]: Fixed

File: alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/src/test/java/org/alfresco/transformer/ImageMagickControllerTest.java (hunk 7 above)
Comment (PunchedChimera) [reply alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/src/test/java/org/alfresco/transformer/ImageMagickControllerTest.java:50]: Added comment

File: alfresco-transformer-base/src/test/java/org/alfresco/transformer/AbstractTransformerControllerTest.java (hunk 8 above)
Comment (PunchedChimera) [reply alfresco-transformer-base/src/test/java/org/alfresco/transformer/AbstractTransformerControllerTest.java:200]: Added comments

File: alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/src/test/java/org/alfresco/transformer/ImageMagickControllerTestBase.java [hunk 11]
```diff
@@ -0,0 +1,389 @@
+/*
+ * #%L
+ * Alfresco Transform Core
+ * %%
+ * Copyright (C) 2005 - 2019 Alfresco Software Limited
```
[the remark this answers is not in the document]
Comment (PunchedChimera) [reply alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/src/test/java/org/alfresco/transformer/ImageMagickControllerTestBase.java]: Updated

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/pom.xml [hunk 12]
```diff
@@ -1,291 +1,299 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (PunchedChimera) [reply alfresco-transform-core-aio/alfresco-transform-core-aio-boot/pom.xml:1]: Not sure why it's flagging the line changes. I added two separate commit, one changing the file to CRLF and another back to LF and neither fixed the compare. Added lines are 41-48. No other changes in the file. 
It looks like the release plugin did something to the line endings ee12dfe8539fc7fdb20d4ff2981f670ad4f25dd3

[the remark this answers is not in the document]
Comment (montgolfiere) [reply alfresco-transform-core-aio/alfresco-transform-core-aio-boot/pom.xml:1]: Still seeing large diff in GitHub PR review ... maybe revert and apply the 8 line change using plain text editor ?

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/main/java/org/alfresco/transformer/AIOController.java (hunk 6 above)
Comment (montgolfiere) [reply alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/main/java/org/alfresco/transformer/AIOController.java:129]: Ok, thanks. 

As an aside, I'm not sure this should be the minimum param now for a well-defined T-API.

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerImageMagickTest.java [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+    {
+        // Ignore the test in super class as the way the AIO transformer provides config is fundementally different.
+
+    }
+    
+    @Test
+    @Override
+    public void deprecatedCommandOptionsTest() throws Exception
+    {
+        // Ignore test in super class as the CommandOptions is not within the imagemagick_engine_config.json
```
Comment (eknizat) [inline alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerImageMagickTest.java:147]: This was added recently as part of https://github.com/Alfresco/alfresco-transform-core/commit/57e4ea4381ae016c453106bd79f53be003977624

File: alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/pom.xml [hunk 14]
```diff
@@ -81,19 +81,23 @@
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-surefire-plugin</artifactId>
-                <configuration>
```
Comment (eknizat) [inline alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/pom.xml:84]: Is this getting removed intentionally?

Comment (PunchedChimera) [reply alfresco-transform-imagemagick/alfresco-transform-imagemagick-boot/pom.xml:84]: Yes this is now moved to the super pom so that it can be accessed by both core-aio and imagemagick. This configuration is only used for test purposes and is not used in the final JARs.

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerImageMagickTest.java (hunk 13 above)
Comment (PunchedChimera) [reply alfresco-transform-core-aio/alfresco-transform-core-aio-boot/src/test/java/org/alfresco/transformer/AIOControllerImageMagickTest.java]: Override has now been removed and inherited test is run.

File: alfresco-transform-core-aio/alfresco-transform-core-aio-boot/pom.xml (hunk 12 above)
[the remark this answers is not in the document]
Comment (PunchedChimera) [reply alfresco-transform-core-aio/alfresco-transform-core-aio-boot/pom.xml]: Fixed, finally....
